developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,360 @@
+3. SOF audio topology for both host and guest.
+4. SOF supporting code in QEMU.
+
+This driver architecture allows full SOF audio use cases to reside on either
+guest VM, or host, or both. It also allows non audio DSP firmware features to
+also be virtualised on guests.
+
+VirtIO_ is a platform-agnostic virtualisation standard, managed by OASIS_. It
+defines a design and a protocol for implementing device IO in VMs.
+
+There are different machine-level virtualisation designs, implementing different
+software configurations. Typically they will have one instance of a general
+purpose OS, running either directly on the hardware, or at least with direct
+unrestricted access to the hardware. This instance is usually called a "host."
+Within that host OS, there is usually one or more user-space processes,
+emulating further independent virtual machines, running their own complete
+instances of general-purpose OSes. Those instances are called "guests."
+
+When Linux is running in a VM, multiple drivers are already available, using
+virtio: networking, storage, etc. Those drivers use the virtio and virtqueue
+APIs to communicate with the host or, potentially, with other guests.
+
+The host part depends on the used virtualisation solution. Some of those
+solutions implement the VirtIO API themselves, then using their method of choice
+to talk to the host they are running on. A typical Linux virtualisation
+implementation uses QEMU + KVM to run guest instances. In this case there are
+two ways to implement a VirtIO host: in the user space and in the kernel. If
+implemented in the user space, it is managed by QEMU. In the kernel space VirtIO
+host drivers use the Linux vhost API.
+
+It is also worth to note, that VirtIO guest drivers reside in respective driver
+subdirectories in the Linux kernel, e.g. drivers/net, drivers/block etc. On the
+contrary, all vhost drivers are located under drivers/vhost. That directory also
+contains vhost.c and vhost.h which are used by all vhost drivers.
+
+Every VirtIO driver requires explicit support in QEMU, which can be very minimal
+and simple in cases where all the functionality is implemented in VirtIO and
+vhost drivers, or it can actually implement a user-space counterpart to VirtIO
+drivers on the guest. In our case the SOF QEMU support code is rather minimal,
+it only adds support for guest reset reporting and guest identification to a set
+of standard and mostly empty device node and virtual queue management
+operations.
+
+2. SOF VirtIO design
+********************
+
+The SOF_ VirtIO design implements an interface to transport IPC (Inter-Process
+Communication) commands and audio data between host and guests. It re-uses the
+SOF-defined IPC API, also used by the native SOF driver to communicate with the
+SOF firmware, running on the audio DSP. This means, that the virtualisation is
+"applied" between the SOF driver and the DSP, i.e. IPC messages that the guest
```

Comment:
virtualization mechanisms are inserted between the SOF guest driver and the DSP hardware.

Reply:
mmh, between the host driver and the DSP? As far as I understand what we are saying is, that the normal case is no virtualisation. Just a (host) SOF driver and a DSP, communicating over IPC. Now we take that interface and bring the "top" part to the guest. So, we "insert" those mechanisms into the original (host) driver?

Reply:
oh, I see what you're saying now.

I don't really have a better wording so let's keep it as is.